src/SecretsFile.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE FlexibleContexts #-}
+
+module SecretsFile where
+
+import Data.Char              (toUpper)
+import Data.Monoid ((<>))
+import Data.List              (intercalate)
+import Control.Monad.Except   (MonadError, throwError)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+
+import qualified Control.Exception          as Exception
+
+import Text.ParserCombinators.ReadP
+
+data Secret = Secret
+  { sMount   :: String
+  , sPath    :: String
+  , sKey     :: String
+  , sVarName :: String
+  } deriving (Eq, Show)
+
+data SecretFileErr
+  = IOError           FilePath
+  | ParseError        FilePath
+
+instance Show SecretFileErr where
+  show err =
+    case err of
+      (IOError fp) ->
+        "An I/O error happened while opening: " <> fp
+      (ParseError fp) ->
+        "Could not parse: " <> fp
+
+data SecretFileVersion = V1 | V2
+
+-- | Run a ReadP parser
+--
+-- Report a syntax error (without additional information) or return a list of
+-- secrets. The way to run a ReadP parser is to use the readP_to_S function and
+-- pass it a string.
+runParser :: MonadError SecretFileErr m => FilePath -> ReadP a -> String -> m a
+runParser fileName parser contents =
+  let result = (readP_to_S parser) contents
+  in case result of
+    [] -> throwError $ ParseError fileName
+    _ -> pure . fst . head $ result
+
+-- | Parser for the entire secret file format
+--
+-- Determines the version of the format, by looking for @VERSION 2@. Default
+-- to V1 if this is absent. The @V1@ format consists of single lines of secret
+-- specifications. The @V2@ format contains multiple "blocks" of secrets. Each
+-- block specifies which mount point it pertains to.
+secretFileP :: ReadP [Secret]
+secretFileP = do
+  fileVersion <- option V1 (formatVersionP <* skipSpaces)
+  case fileVersion of
+    V1 -> many ((secretP fileVersion "secret") <* skipSpaces) <* eof
+    V2 -> concat <$> many (secretBlockP <* skipSpaces) <* eof
+
+formatVersionP :: ReadP SecretFileVersion
+formatVersionP = const V2 <$> string "VERSION 2"
```

Comment:
Should `VERSION 2` be on its own line? Do you want to strip leading whitespace as well here?